rust/gui-client/src-tauri/src/client/ipc/windows.rs

```diff
@@ -0,0 +1,111 @@
+use anyhow::{anyhow, Context as _, Result};
+use connlib_client_shared::Callbacks;
+use firezone_headless_client::{IpcClientMsg, IpcServerMsg};
+use futures::{SinkExt, Stream};
+use secrecy::{ExposeSecret, SecretString};
+use std::{pin::pin, task::Poll};
+use tokio::{net::windows::named_pipe, sync::mpsc};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+pub(crate) struct Client {
+    task: tokio::task::JoinHandle<Result<()>>,
+    // Needed temporarily to avoid a big refactor. We can remove this in the future.
+    tx: mpsc::Sender<String>,
+}
+
+impl Client {
+    pub(crate) async fn disconnect(mut self) -> Result<()> {
+        self.send_msg(&IpcClientMsg::Disconnect)
+            .await
+            .context("Couldn't send Disconnect")?;
+        self.task.abort();
+        Ok(())
+    }
+
+    #[allow(clippy::unused_async)]
+    pub(crate) async fn send_msg(&mut self, msg: &IpcClientMsg) -> Result<()> {
+        self.tx
+            .send(serde_json::to_string(msg).context("Couldn't encode IPC message as JSON")?)
+            .await
+            .context("Couldn't send IPC message")?;
+        Ok(())
+    }
+
+    pub(crate) async fn connect(
+        api_url: &str,
+        token: SecretString,
+        callback_handler: super::CallbackHandler,
+        tokio_handle: tokio::runtime::Handle,
+    ) -> Result<Self> {
+        tracing::info!(pid = std::process::id(), "Connecting to IPC service...");
+        let ipc = named_pipe::ClientOptions::new()
+            .open(firezone_headless_client::windows::pipe_path())
+            .context("Couldn't connect to named pipe server")?;
+        let ipc = Framed::new(ipc, LengthDelimitedCodec::new());
+        // This channel allows us to communicate with the GUI even though NamedPipeClient
+        // doesn't have `into_split`.
+        let (tx, mut rx) = mpsc::channel(1);
+
+        let task = tokio_handle.spawn(async move {
+            let mut ipc = pin!(ipc);
+            loop {
```

Comment:
I think a better way to solve this is to:

1. Split `ipc` into a `ReadHalf` and a `WriteHalf`: https://docs.rs/futures/latest/futures/io/trait.AsyncReadExt.html#method.split
2. Construct separate `FramedRead` and `FramedWrite` objects.
3. Have dedicated loops using `async/await` where you forward from one to the other. You might even be able to use `forward`: https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.forward to take elements from the channel and forward them to the `ipc`.

Reply:
Shoot, I didn't think that was possible since I'd only seen `into_split` on the Unix domain sockets.
Yeah this will be a lot simpler. I'll do it next week in refactoring.
I thought some objects just couldn't be split lol

Does it create a mutex internally to handle AsyncRead and AsyncWrite both being `&mut`?

Reply:
> Does it create a mutex internally to handle AsyncRead and AsyncWrite both being `&mut`?

It uses a thing called a `BiLock` (https://docs.rs/futures-util/0.3.30/src/futures_util/lock/bilock.rs.html) that is more optimised for the special-case of only two owners.

Reply:
Ended up using `tokio::io::split` and that's on `main` after a recent refactor landed